### PR TITLE
Remove old VimeoClient and related code in vimeo-networking

### PR DESCRIPTION
**Pull request by anthonycr**
# Summary
- Removed all code from `vimeo-networking` module. This module now simply aggregates `api-core`, `auth`, and `request` modules together. Consumers can rely on the modules individually, or they can use `vimeo-networking` like before. Now they are required to specify a `model-{type}` implementation. If they don't care about serialization, then they can just use `models`. If they are java consumers, then `models-serializable` might be useful, and if android consumers `models-parcelable`.

- Rewrote example. Example now showcases configuration of the `Authenticator` and `VimeoApiClient`, authentication via client credentials and code grant, and fetching the current user's videos. Example now resides in `example` and is only written in Kotlin.

- Updated the models plugin to only register tasks once. This fixes an issue with compiling the `models-parcelable` module.

- Renamed `Connection` to `BasicConnection`, and created a common `Connection` interface that can be shared by all connections, similar to the way `Interaction` works.

- Exposed moshi configuration in `RetrofitSetupModule`.

- Removed `example-android` and `example-java` that used old `VimeoClient`.

- Added support in `Authenticator` for
  - Log in via pin code.
  - Log in via code grant.
  - Exchanging one access token for another.

### CODE REVIEW

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt:168]: Some line comments here would be helpful to tell the user where/how to get the client id and secret.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt:97]: It might also be helpful to point them at the example since that demonstrates how to authenticate with the client: `example/src/main/java/com/vimeo/example/MainActivity.kt`

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt:128]: Yes! Great to see this annotation in action. Do you think we should ticket an audit of the networking library to tag any outstanding internals or do you feel confident that you tagged all of them?

File: example/src/main/res/values/strings.xml
Comment (howardrigberg) [inline example/src/main/res/values/strings.xml]: Maybe something like: "Vimeo Networking Authorization Example"

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: It would be a little more helpful and self documenting to handle `it.reason.errorCode == EMPTY_AUTHENTICATION` and give a warning to the user asking if they have set their `CLIENT_ID` and `CLIENT_SECRET`.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: It would be clearer to have a named static for this near the `CLIENT_ID` and `CLIENT_SECRET` with line comments that explain why and how it will be used.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: Might want to indicate that it is ok to call this on the same thread. No network requests are made.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: I am confused about how this would ever get called with the current code. The browser is launched and then the user will be navigated to the redirect url that they created in the web console, but the app won't get called back with the redirectUri. Is there something I am missing here? 

If we don't want to deal with the complications around managing the extra code that would be required to do webview based auth, we could give instructions for the user to copy paste the url into a edittext manually and click a button to complete the flow manually. It is an ugly approach, but it gets the job done for demo purposes with a minimal amount of work. It could come along with line comments about how you might actually handle it for real in mobile code.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt:77]: It would be a little more work, but it would be nice to have a single login/logout button that changed purpose depending on the logged in state.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: Maybe we could add a button that would hit staff picks or something else that required client credentials after successful authentication? It would be nice to illustrate the next step. Also, it might be illustrative to label that button as "use client credentials" rather than "get staff picks" because the focus here is on the authorization flow and that will act as a final confirmation step in the process. It also might help to put the related buttons for authorization code grant and client credentials into separate visual groups in the demo app.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt:96]: Alternatively, it might help to label the button as "use authorization code grant" rather than "get my videos" since the point is to validate and demonstrate the last step in the authorization process.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: It would be helpful and illustrative to actively report when the user is logged out since you will know it at this point rather than blindly making the network request.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: Would be helpful to explain where and how the request code is used in a line comment either here or near `obtainCodeGrantAuthorizationUri` above. (i.e. The REQUEST_CODE will be passed as a `state` query string parameter to the redirect uri  that is provided in `withCodeGrantRedirectUrl` when logging in the user with code grant authorization. It can be used for security validation by the receiving code.)

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: Maybe explain the flow a little more here: You need to open the uri in a browser or webview and then pass the entire resultant redirect uri to this method.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt]: Why don't we just remove `logInWithFacebook` and `logInWithGoogle` and rename the other existing `joinWith...` methods as `loginOrJoinWithFacebook` and `loginOrJoinWithGoogle`? The `logInWith...` methods are internal and unused and it probably just confusing to keep them around at all at this point.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt]: An arbitrary value. We return this value to you later as a security measure in a query string parameter named `state` in the callback to the [redirectUri] you specified above.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt:202]: The uri needs to be preregistered at https://developer.vimeo.com/apps/ for your developer accoount with Vimeo.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt:201]: The client identifier of your app. To get this value, access your app's information page from https://developer.vimeo.com/apps/.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (WhosNickDoglio) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: Just curious but why isn't this method or the related ones using `VimeoRequest` and `VimeoCallback` ?

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (WhosNickDoglio) [inline example/src/main/java/com/vimeo/example/MainActivity.kt:77]: I'm getting an error when I try this. 
Screenshot_1597088056

**Review by WhosNickDoglio — APPROVED** [review]
This looks great! 

Excited to have this fully in Kotlin. 

Left a question and an issue I noticed with the example 👍

File: vimeo-networking/src/main/java/com/vimeo/networking/AccountStore.java
Comment (howardrigberg) [inline vimeo-networking/src/main/java/com/vimeo/networking/AccountStore.java:1]: You left vimeo-networking/src/main/resources/keystore.bks. Is this intentional?

**Review by howardrigberg — CHANGES_REQUESTED** [review]
Nice job, I've got a few line comments. Mostly on the example.

Also, wondering how you feel about changing `com.vimeo.networking2` to `com.vimeo.networking` now that all of the old stuff is deleted.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt:77]: Not sure what changes you made to the example before running it, but because the docs aren't clear - you need to create an API app here: https://developer.vimeo.com/apps/new and then register the redirect URI (in the example it's set to `https://example.com`) on the API page - navigate to https://developer.vimeo.com/apps and then choose your app and at the bottom click "add URL." Once this is done replace the client id and client secret placeholders in the example.

**Comment by anthonycr** [comment]
Regarding changing the package name, we made the decision to use the package name `networking2` in order to facilitate migration to the new networking library, as this is a complete rewrite of the library. Since we aren't the only consumers, and everyone else is still on V1, it is better to leave it as is.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: Well this function is just constructing a URL so it doesn't enqueue a request or require a callback.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: The app fires off an intent to open the code grant URL in the browser. The website prompts the user to log in and authorize the app to access Vimeo. Once that is done, the browser navigates to the redirect URI specified in the configuration. The application has to register that URI in the manifest (see `AndroidManifest.xml` where `https://example.com` is registered in an intent filter). When the browser redirects to `https://example.com?code={auth code}` the app is reopened and the redirect URI is present in the `intent`. The application then extract the URI from the `intent` and pass it to `authenticateWithCodeGrant`, which extracts the authorization code and authenticates with the API. Since `MainActivity` is restarted when a new intent is fired, `onCreate` is called. If `MainActivity` was brought to the front instead, this code would need to be in `onNewIntent`.

The 3 things the user has to do is
1. Set the client ID in this file.
2. Set the client secret in this file.
3. Add `https://example.com` as a callback URL in the API dashboard.

Once those three things are done, this code works as expected.

Comment (howardrigberg) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: Ah. I see. Thanks for the explanation. I thought I might be missing something. That's definitely something that needs to be called out clearly in the docs. Did you test the entire flow to confirm that it works?

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: I changed the name to `createCodeGrantAuthorizationUri` and updated the docs a little. It feels a bit weird to call out this method as being thread safe when all the functions on `Authenticator` are thread safe and we don't call those out.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt:77]: I want to keep the example as simple as possible.

File: example/src/main/res/values/strings.xml
Comment (anthonycr) [reply example/src/main/res/values/strings.xml]: I'll change it to just `Vimeo Networking Example`. It also demonstrates how to make a request with the `VimeoApiClient` so it encompasses a little more than auth.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt:168]: Added documentation to the class definition instead that explains setting up the project.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (WhosNickDoglio) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: Ah, I saw the use of `authService` in `AuthenticatorImpl` and assumed it was a call, didn't notice the `request().url().toString()`, makes sense now 👍

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt:201]: This is an internal class, so this function doesn't need comprehensive docs as if it was publicly consumable. I added a link to the documentation URL to the `VimeoApiConfiguration` constructor and builder just to be safe.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: FWIW, Nick's confusion around this was shared. Most of what the Authenticator does is make requests to our servers. This method is the exception. When I first saw it in action in the example, I wasn't immediately sure if the source of the string was from the server or not or why it was different. Once I looked at the implementation, I understood what was happening, but I was initially a little confused. That's why I made the thread suggestion initially. If you have any other ideas of docs that could call this out in some way, it would be helpful. Completely optional though.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt:202]: This is an internal class not accessible to the consumer so this documentation isn't necessary. There is already public documentation about this requirement on `VimeoApiConfiguration.codeGrantRedirectUri`.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (howardrigberg) [reply auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt:201]: Perfect! That is what I was literally about to suggest.

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt]: This is an internal class that is not visible to the consumer, so I added this documentation to the `Authenticator` instead.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: Yes I created my own API app that I used to verify this example while building it. This is the same flow supported by the V1 library.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (WhosNickDoglio) [reply example/src/main/java/com/vimeo/example/MainActivity.kt:77]: I was setting up the code grant URL wrong, updated docs helped me get it working properly. 👍

File: auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/internal/AuthService.kt]: I want to keep them around as documentation of their existence and how they can be used in contrast with the `joinWith...` functions. Since `AuthService` is internal, this will not be confusing to the user, only us as maintainers. I added some more docs to these functions to explain that they are intentionally unused and added suppression of the unused warnings.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt:97]: Instead of linking to the `MainActivity` which is a bit awkward since the example isn't packaged with the library, I added more to the `Authenticator` documentation block about client credentials vs basic authentication vs code grant. We will be linking to the example in the readme as well.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt:128]: They definitely aren't all tagged, it's something we can tackle as we go along. I finished adding `@Internal` annotations to the `Authenticator` in a later PR, but we don't have any in the `VimeoApiClient`. I think that's fine for now as the user will be able to see in the response from the server that they don't have access to the content.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: Now that the `MainActivity` explicitly calls this out in the documentation at the top, this is no longer necessary?

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: I added a staff picks fetch and removed the client credentials button. This brings the example more in line with the original, where client credentials are implicitly requested at startup rather than by clicking a button. Now if the user clicks on the get my videos button without being logged in it will set the message to say to log in. However, for staff picks, the consumer will see that they can make the request without being logged in.

Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: Overall, I want to keep the example as short as possible. The current example is less than 200 lines of code, and my goal is to keep the `MainActivity` under that. The longer the example and the more complicated it is, the harder it will be for users to grok.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt:96]: Discussed on the staff picks comment.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: I think it is still a helpful thing to have. People will make mistakes regardless of the docs (many never read them) and this would be a quick way to point them in the right direction.

**Comment by anthonycr** [comment]
@howardrigberg @WhosNickDoglio thanks for the review, I improved the example and documentation so that configuring the example and using the `Authenticator` to perform a code grant is hopefully more clear now.

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (anthonycr) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: Since this isn't the only request that can fail due to the id and secret not being set, I am just going to add a dialog at the beginning of `onCreate` that stops them in their tracks.

Comment (howardrigberg) [reply example/src/main/java/com/vimeo/example/MainActivity.kt]: Great idea!

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt:54]: Example doesn't match description above. `invoke` is not... ahem... invoked.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt]: You could mention that in practice the credentials are automatically stored in the AccountStore associated with the Authenticator. It might provide some clarity about the magic at play with how the credentials are retained.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt:94]: It would be helpful to also mention the need to register an `intent-filter` for the redirect uri in the manifest so your app will get called back.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (howardrigberg) [inline auth/src/main/java/com/vimeo/networking2/Authenticator.kt:240]: and also a `state` query parameter that matches the response code set in createCodeGrantAuthorizationUri that can be used to validate that uri originated from your app

File: example/src/main/res/layout/activity_main.xml
Comment (howardrigberg) [inline example/src/main/res/layout/activity_main.xml]: "Log In"

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt]: Can you show it as a toast instead so it is consistent with the other errors?

File: example/src/main/java/com/vimeo/example/MainActivity.kt
Comment (howardrigberg) [inline example/src/main/java/com/vimeo/example/MainActivity.kt:49]: This is perfect!

**Review by howardrigberg — APPROVED** [review]
Looks good, Anthony. I like the documentation improvements. Things are much clearer now. I made a few more suggestions to docs and code. Feel free to adopt as many as you agree with. I'm approving it now though.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt:54]: `Authenticator(Configuration)` calls `Authenticator.invoke(Configuration)` in Kotlin, it's syntactic sugar. The `invoke` function is an operator function like `plus` or `minus`.

File: auth/src/main/java/com/vimeo/networking2/Authenticator.kt
Comment (anthonycr) [reply auth/src/main/java/com/vimeo/networking2/Authenticator.kt:240]: While it's true that the `state` parameter will be present in the redirect URI, we don't validate it for the consumer and leave it up to them to validate. Adding it as a requirement in the param doc is a bit misleading.

**Pull request merged**
